### ISSUE TITLE
Using state.set(...) leads to a leak

### ISSUE THREAD

**Issue by Artur-Galstyan** [patrick-kidger/equinox#586]
Hey Patrick,
I've got another tricky bug, I'm having trouble debugging. I found it while testing the `RoPEEmbeddings` PR and managed to break it down to a minimal code example.

Here's the MVC:

```python
import equinox as eqx
import jax.lax
import jax.numpy as jnp


class Stateful(eqx.Module):
    kv_cache_index: eqx.nn.StateIndex

    max_seq_len: int = eqx.field(static=True)
    embedding_dim: int = eqx.field(static=True)
    num_heads: int = eqx.field(static=True)

    def __init__(self, max_seq_len: int, num_heads: int, embedding_dim: int):
        self.kv_cache_index = eqx.nn.StateIndex(
            init=(
                jnp.zeros(shape=(max_seq_len, num_heads, embedding_dim)),  # key
                jnp.zeros(shape=(max_seq_len, num_heads, embedding_dim)),  # value
                0,  # index
            )
        )
        self.max_seq_len = max_seq_len
        self.embedding_dim = embedding_dim
        self.num_heads = num_heads

    def __call__(self, x, state: eqx.nn.State):
        key, value, index = state.get(self.kv_cache_index)

        # error happens here
        # key = jax.lax.dynamic_update_slice_in_dim(
        #     key,
        #     x,
        #     index,
        #     axis=0,
        # )
        # value = jax.lax.dynamic_update_slice_in_dim(
        #     value,
        #     x,
        #     index,
        #     axis=0,
        # )
        # uncomment this to not get the error?
        state = state.set(self.kv_cache_index, (key, value, index + 1))

        # this works too
        # state = state.set(
        #     self.kv_cache_index, (jnp.ones_like(key), jnp.ones_like(value), index + 1)
        # )
        return x, state


def test_stateful():
    batch_size = 64
    max_seq_len = 16
    num_heads = 8
    embedding_dim = 4
    x = jnp.ones(shape=(batch_size, 1, num_heads, embedding_dim))
    model, state = eqx.nn.make_with_state(Stateful)(
        max_seq_len, num_heads, embedding_dim
    )

    non_batched_x = jnp.zeros(shape=(1, num_heads, embedding_dim))
    non_batched_x, state = model(non_batched_x, state)
    print("So far, so good")
    x, state = eqx.filter_vmap(model, in_axes=(0, None), out_axes=(0, None))(x, state)
    x, state = eqx.filter_vmap(model, in_axes=(0, None), out_axes=(0, None))(x, state)


if __name__ == "__main__":
    with jax.checking_leaks():
        test_stateful()
```

Here are my findings: 

1: Uncomment this part to get the leak:
```python
        key = jax.lax.dynamic_update_slice_in_dim(
            key,
            x,
            index,
            axis=0,
        )
        value = jax.lax.dynamic_update_slice_in_dim(
            value,
            x,
            index,
            axis=0,
        )
```
This one says:
```
This BatchTracer with object id 4845592288 was created on line:
  /Users/arturgalstyan/Workspace/kira/kira/state_bug_3.py:29:14 (__call__) <-- this is the key = jax.lax... part
<BatchTracer 4845592288> is referred to by <list 4846304256>[0]
<list 4846304256> is referred to by <tuple 4846103040>[0]
<tuple 4846103040> is referred to by <_FlattenedData 4844482320>.static_field_values
```

2: Change the "State setting part" to this to *not* get the leak
```python
# This does not leak!
key = jax.lax.dynamic_update_slice_in_dim(
    key,
    x,
    index,
    axis=0,
)
value = jax.lax.dynamic_update_slice_in_dim(
    value,
    x,
    index,
    axis=0,
)
# state = state.set(self.kv_cache_index, (key, value, index + 1))

state = state.set(
    self.kv_cache_index, (jnp.ones_like(key), jnp.ones_like(value), index + 1)
)
```

3: This does also not leak:
```python
key, value, index = state.get(self.kv_cache_index)
# no leaks here!
# key = jax.lax.dynamic_update_slice_in_dim(
#     key,
#     x,
#     index,
#     axis=0,
# )
# value = jax.lax.dynamic_update_slice_in_dim(
#     value,
#     x,
#     index,
#     axis=0,
# )

state = state.set(self.kv_cache_index, (key, value, index + 1))
```

4: Here I just tried something different, but still the leak:
```python
# error happens here
other_key = jax.lax.stop_gradient(key)
other_key = jax.lax.dynamic_update_slice_in_dim(
    other_key,
    x,
    index,
    axis=0,
)
# value = jax.lax.dynamic_update_slice_in_dim(
#     value,
#     x,
#     index,
#     axis=0,
# )
state = state.set(self.kv_cache_index, (other_key, value + 1, index + 1))
```

5: strangely, this does not leak:

```python
# key = jax.lax.dynamic_update_slice_in_dim(
#     key,
#     x,
#     index,
#     axis=0,
# )
# value = jax.lax.dynamic_update_slice_in_dim(
#     value,
#     x,
#     index,
#     axis=0,
# )
key += jnp.ones_like(key)
state = state.set(
    self.kv_cache_index, (key + jnp.ones_like(key), value + 1, index + 1)
)
```

It seems that as soon as I do something with `key` using `jax.lax`, the leak happens. Do you perhaps have an idea?

In general, the Jax docs say that this error only happens, as soon as you change a global state outside a transformation (such as the VMAP in the beginning). My only guess would be that it has something to do both with the `jax.lax...` and setting the state using `state.set(...)`.

**Comment by patrick-kidger** [comment]
Your `out_axes` is misspecified. The output `state` should be batched (as it interacts with a batched `x` within your model.) Instead of `out_axes=(0, None)` you want `out_axes=(0, 0)`. (Which is equivalent to `out_axes=0`, which is also the default.)

**Comment by Artur-Galstyan** [comment]
Alright, I settled on this:

```python
x, state = eqx.filter_vmap(model, in_axes=(0, None))(x, state)
# output shapes are:
# (64, 1, 8, 4)
# state: State(0x105bb2e98=(f32[64,16,8,4], f32[64,16,8,4], i32[64]))
x, state = eqx.filter_vmap(model, in_axes=(0, 0))(x, state)
```

When not defining the `out_axis` a new batch axis gets attached (the number `64`). In subsequent calls, I define `in_axis=(0, 0)` to indicate the new batch axis in the state. The only weird thing is that the `index` integer field is duplicated `batch_size` times 🤔 

Is this the intended way of using the states with `vmap`? (I'm still very new to the State API; still practicing 😄)

**Comment by patrick-kidger** [comment]
Yup, I think this is now all correct / expected!

**Comment by Artur-Galstyan** [comment]
Okay, I went one step further and I think I found a good way to work with models with states. 

First, I create the model and the batched state first:

```python
def batch_make_with_state(batch):
    return eqx.nn.make_with_state(Model)(
        ...
    )


model, states = eqx.filter_vmap(batch_make_with_state, out_axis=(None, 0))(jnp.zeros(batch_size))
```

This gives me the model, as well as the batched states:

(64 is the batch size in my case)
```
states: State(0x102fbee98=(f32[64,32,6,384], f32[64,32,6,384], i32[64,1]))
```

Then I can use the `vmap` without changing the `in_axis` between the first and second call:

```python
# first call
output, states = eqx.filter_vmap(model)(x, states)

# second call
output, states = eqx.filter_vmap(model)(x, states)
```

Thanks for the help!

**Comment by Artur-Galstyan** [comment]
Yes, that also worked like a charm! Closing this then.
